src/Aspire.Hosting.Python/PythonAppResourceBuilderExtensions.cs

```diff
@@ -256,6 +258,48 @@ private static IResourceBuilder<PythonAppResource> AddPythonAppCore(
             // This will set up the the entrypoint based on the PythonEntrypointAnnotation
             .WithEntrypoint(entrypointType, entrypoint);
 
+        // Add virtual environment validation before resource starts
+        resourceBuilder.OnBeforeResourceStarted(async (resource, evt, ct) =>
+        {
+            // Only check if interaction service is available (dashboard is enabled)
+            var interactionService = evt.Services.GetService<IInteractionService>();
+            if (interactionService is null || !interactionService.IsAvailable)
+            {
+                return;
+            }
+
+            // Get the virtual environment path from the annotation
+            if (!resource.TryGetLastAnnotation<PythonEnvironmentAnnotation>(out var pythonEnv) ||
+                pythonEnv.VirtualEnvironment is null)
+            {
+                return;
+            }
+
+            // Skip check if UV is being used (it creates the virtual environment automatically)
+            if (pythonEnv.Uv)
+            {
+                return;
+            }
+
+            var venvPath = pythonEnv.VirtualEnvironment.Location;
+
+            // Check if the virtual environment directory exists
+            if (!Directory.Exists(venvPath))
+            {
+                var title = "Python virtual environment not found";
+                var message = $"The Python virtual environment for resource '{resource.Name}' was not found at '{venvPath}'. Please create the virtual environment before running the application.";
+                var options = new NotificationInteractionOptions
+                {
+                    Intent = MessageIntent.Error,
+                    LinkText = "Learn more",
+                    LinkUrl = "https://aka.ms/aspire/python"
+                };
+
+                // Show notification without blocking (fire and forget)
+                _ = interactionService.PromptNotificationAsync(title, message, options, ct);
```

Comment:
@davidfowl This will create a new prompt every time the resource is started even if a prompt is already visible. If there are multiple Python resources in this state, you'll get a separate prompt for each. It seems we really want this coalescing notification capability in the core so all these integrations don't need to keep implementing it?